### PR TITLE
[Feature] support integrate starmgr

### PR DESCRIPTION
**Pull request by starrocks-xupeng**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 20 / 238 (08.40%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|6|00.00%|[350, 351, 352, 353, 354, 356]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|0|93|00.00%|[32, 40, 41, 42, 44, 46, 48, 50, 51, 54, 55, 56, 57, 60, 64, 65, 70, 71, 73, 74, 75, 76, 77, 82, 84, 85, 86, 87, 89, 91, 92, 93, 94, 95, 96, 97, 101, 102, 103, 108, 109, 110, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 126, 127, 130, 132, 136, 137, 138, 139, 140, 144, 146, 147, 148, 149, 150, 151, 152, 153, 155, 156, 159, 163, 164, 166, 167, 170, 171, 174, 175, 177, 178, 179, 180, 181, 184, 185, 186, 188, 192]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 231, 232, 234, 235, 236, 237, 240]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[942, 943, 944, 1595, 1596]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[709, 710, 711]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|8|96|08.33%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 68, 69, 72, 73, 77, 78, 80, 83, 84, 85, 86, 98, 102, 106, 107, 109, 110, 111, 112, 114, 116, 117, 118, 122, 123, 124, 125, 126, 127, 128, 131, 132, 135, 136, 139, 142, 144, 145, 146, 147, 150, 151, 154, 155, 156, 158, 159, 161, 163, 164, 165, 167, 169, 173, 176, 177, 178, 179, 180, 185, 186, 187, 188, 189, 190, 191, 193, 194, 197, 198, 201, 202, 203, 204, 206, 209, 213, 217]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 72 / 238 (30.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|6|00.00%|[350, 351, 352, 353, 354, 356]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 231, 232, 234, 235, 236, 237, 240]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[942, 943, 944, 1595, 1596]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[709, 710, 711]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|8|96|08.33%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 68, 69, 72, 73, 77, 78, 80, 83, 84, 85, 86, 98, 102, 106, 107, 109, 110, 111, 112, 114, 116, 117, 118, 122, 123, 124, 125, 126, 127, 128, 131, 132, 135, 136, 139, 142, 144, 145, 146, 147, 150, 151, 154, 155, 156, 158, 159, 161, 163, 164, 165, 167, 169, 173, 176, 177, 178, 179, 180, 185, 186, 187, 188, 189, 190, 191, 193, 194, 197, 198, 201, 202, 203, 204, 206, 209, 213, 217]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|52|93|55.91%|[74, 75, 92, 93, 94, 95, 108, 109, 110, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 126, 127, 136, 137, 138, 139, 140, 148, 149, 150, 151, 152, 153, 174, 175, 177, 178, 181, 185, 186]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 121 / 235 (51.49%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|6|00.00%|[350, 351, 352, 353, 354, 356]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 231, 232, 234, 235, 236, 237, 240]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[931, 932, 933, 1576, 1577]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[703, 704, 705]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|50|93|53.76%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 72, 73, 77, 78, 83, 84, 85, 86, 97, 106, 107, 112, 113, 134, 137, 139, 140, 141, 142, 145, 146, 159, 160, 173, 174, 175, 185, 186, 189, 198, 199, 201, 208]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|59|93|63.44%|[74, 75, 92, 93, 94, 95, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 136, 137, 138, 139, 140, 148, 149, 150, 151, 152, 153, 174, 175, 177, 178, 181, 185, 186]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Review by tracymacding — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/staros/StarMgrServer.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+        // 3. write new image
+        // Write image.ckpt
+        Storage storage = new Storage(imageDir);
+        File imageFile = storage.getImageFile(getJournalSystem().getReplayId());
+        File ckpt = new File(imageDir, Storage.IMAGE_NEW);
+        if (!ckpt.exists()) {
+            if (!ckpt.getParentFile().exists()) {
+                LOG.info("create image dir for star mgr, {}.", ckpt.getParentFile().getAbsolutePath());
+                ckpt.getParentFile().mkdir();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/staros/StarMgrServer.java:186]: Check return value.

File: fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            }
+            replayer = null;
+        }
+
+        // replay all journal and start journal writer
+        try {
+            bdbjeJournal.open();
+
+            long replayStartTime = System.currentTimeMillis();
+            replayTo(-1);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java:85]:
```suggestion
            replayTo(JournalCursor.CUROSR_END_KEY);
```

CUROSR_END_KEY -> CURSOR_END_KEY

File: fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        replayer = null;
+    }
+
+    // for checkpoint thread only
+    public BDBJEJournalSystem(BDBJEJournal journal) {
+        bdbjeJournal = journal;
+        replayedJournalId = new AtomicLong(0L);
+    }
+
+    public long getReplayId() {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java:59]: getReplayedId or getReplayedJournalId are better.

Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java:59]: that needs to release a new star mgr version, maybe later

File: fe/fe-core/src/main/java/com/starrocks/staros/StarMgrJournal.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+package com.starrocks.staros;
+
+import com.staros.journal.Journal;
+import com.starrocks.common.io.Writable;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+// wrapper for star manager journal to help write to dbdje
```
[the remark this answers is not in the document]
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/staros/StarMgrJournal.java]: done

File: fe/fe-core/src/main/java/com/starrocks/server/NodeMgr.java [hunk 5]
```diff
@@ -345,15 +346,14 @@ public void getClusterIdAndRoleOnStartup() throws IOException {
...
-                // if (!dir.exists()) {
-                //     LOG.info("create image dir for {}.", dir.getAbsolutePath());
-                //     dir.mkdir();
-                // }
-                // getNewImageOnStartup(rightHelperNode, StarMgrServer.IMAGE_SUBDIR);
+                String subDir = this.imageDir + StarMgrServer.IMAGE_SUBDIR;
+                File dir = new File(subDir);
+                if (!dir.exists()) { // subDir might not exist
+                    LOG.info("create image dir for {}.", dir.getAbsolutePath());
+                    dir.mkdir();
```
[the remark this answers is not in the document]
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/server/NodeMgr.java]: done

File: fe/fe-core/src/main/java/com/starrocks/staros/StarMgrServer.java (hunk 1 above)
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/staros/StarMgrServer.java]: done

File: fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        }
+    }
+
+    private boolean replayJournal(JournalCursor cursor)
+            throws JournalException, InterruptedException, JournalInconsistentException {
+        long startReplayId = replayedJournalId.get();
+        long startTime = System.currentTimeMillis();
+        while (true) {
+            JournalEntity entity = null;
+            entity = cursor.next();
```
[the remark this answers is not in the document]
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java]: done

File: fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java (hunk 2 above)
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/staros/BDBJEJournalSystem.java]: done

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 7]
```diff
@@ -1584,6 +1584,16 @@ public class Config extends ConfigBase {
     public static String starmgr_address = "127.0.0.1:6090";
     @ConfField
     public static boolean integrate_starmgr = false;
+    @ConfField
+    public static String starmgr_s3_bucket = "";
+    @ConfField
+    public static String starmgr_s3_region = "";
```
[the remark this answers is not in the document]
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java]: removed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 118 / 236 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|8|00.00%|[350, 351, 352, 353, 354, 355, 356, 359]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 231, 232, 234, 235, 236, 237, 240]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[931, 932, 933, 1576, 1577]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[703, 704, 705]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|49|94|52.13%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 72, 73, 77, 78, 83, 84, 85, 86, 97, 106, 107, 112, 113, 133, 136, 138, 139, 140, 141, 144, 145, 158, 159, 172, 173, 174, 184, 185, 186, 187, 191, 200, 201, 203, 210]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|58|92|63.04%|[74, 75, 92, 93, 94, 95, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 136, 137, 138, 139, 140, 148, 149, 150, 151, 152, 153, 173, 174, 176, 177, 180, 184, 185]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 244 (48.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|8|00.00%|[350, 351, 352, 353, 354, 355, 356, 359]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 224, 225, 227, 228, 229, 230, 233]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[931, 932, 933, 1576, 1577]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[703, 704, 705]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|48|91|52.75%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 72, 73, 77, 78, 83, 84, 85, 86, 97, 106, 107, 128, 131, 133, 134, 135, 136, 139, 140, 153, 154, 167, 168, 169, 179, 180, 181, 182, 186, 195, 196, 198, 205]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|60|103|58.25%|[74, 75, 92, 93, 94, 95, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 146, 147, 148, 149, 150, 159, 161, 162, 163, 164, 165, 166, 167, 168, 170, 171, 176, 177, 179, 181, 195, 196, 198, 199, 202, 206, 207]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 122 / 238 (51.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|8|00.00%|[350, 351, 352, 353, 354, 355, 356, 359]|
|:large_blue_circle:|com/starrocks/StarRocksFE.java|0|11|00.00%|[119, 120, 121, 122, 123, 124, 125, 126, 127, 130, 131]|
|:large_blue_circle:|com/starrocks/leader/Checkpoint.java|0|8|00.00%|[95, 224, 225, 227, 228, 229, 230, 233]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|5|00.00%|[931, 932, 933, 1576, 1577]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[703, 704, 705]|
|:large_blue_circle:|com/starrocks/http/meta/MetaService.java|0|2|00.00%|[212, 277]|
|:large_blue_circle:|com/starrocks/staros/StarMgrServer.java|48|91|52.75%|[39, 43, 47, 48, 50, 53, 56, 57, 59, 61, 72, 73, 77, 78, 83, 84, 85, 86, 97, 106, 107, 128, 131, 133, 134, 135, 136, 139, 140, 153, 154, 167, 168, 169, 179, 180, 181, 182, 186, 195, 196, 198, 205]|
|:large_blue_circle:|com/starrocks/staros/BDBJEJournalSystem.java|63|97|64.95%|[74, 75, 92, 93, 94, 95, 115, 116, 117, 118, 119, 120, 122, 123, 124, 125, 146, 147, 148, 149, 150, 158, 159, 160, 161, 162, 163, 183, 184, 186, 187, 190, 194, 195]|
|:large_blue_circle:|com/starrocks/staros/StarMgrJournal.java|6|8|75.00%|[30, 31]|
|:large_blue_circle:|com/starrocks/common/Config.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBJournalCursor.java|1|1|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**